File: jscomp/test/rec_module_opt.js
'use strict';

var $$Set = require("../../lib/js/set.js");
var Caml = require("../../lib/js/caml.js");
var Curry = require("../../lib/js/curry.js");

function compare(t1, t2) {
  if (t1.TAG === /* Leaf */0) {
    if (t2.TAG === /* Leaf */0) {
      return Caml.string_compare(t1._0, t2._0);
    } else {
      return 1;
    }
  } else if (t2.TAG === /* Leaf */0) {
    return -1;
  } else {
    return Curry._2(ASet.compare, t1._0, t2._0);
  }
}

var A = {
  compare: compare
};

var ASet = $$Set.Make(A);

var X0 = {};

var Y0 = {};

function f(x) {
  return x + 1 | 0;
}

var X1 = {
  f: f
};

function f$1(x) {
  return x + 2 | 0;
}

var Y1 = {
  f: f$1
};

var X;

exports.A = A;
exports.ASet = ASet;
exports.X = X;
exports.X0 = X0;
exports.Y0 = Y0;
exports.X1 = X1;
exports.Y1 = Y1;
/* ASet Not a pure module */

File: jscomp/test/rec_module_opt.ml

[@@@bs.config {
  flags = [|
  (* "-drawlambda"; *)
  (* "-dlambda";  *)
  (* "-dtypedtree"; *)
  (* "-bs-diagnose" *)

  "-bs-no-cross-module-opt"
  |]
}]

module rec A : sig
  type t = Leaf of string | Node of ASet.t
  val compare: t -> t -> int
end = struct
  type t = Leaf of string | Node of ASet.t
  let compare t1 t2 =
    match (t1, t2) with
    | (Leaf s1, Leaf s2) -> Pervasives.compare s1 s2
    | (Leaf _, Node _) -> 1
    | (Node _, Leaf _) -> -1
    | (Node n1, Node n2) -> ASet.compare n1 n2
  let hello x = x  
end
and ASet
  : Set.S with type elt = A.t
  = Set.Make(A)

module rec X : sig end = X  


module rec X0 : sig 
  type t   
end   = struct 
  type t  
end   
and Y0 : sig 
  type t   
end   = struct 
  type t   
end    
module type S = sig 
  val f : int -> int    
end   

module rec X1 : S = struct 
  let f x = x + 1
end  
and Y1 : S = struct 
  let f x = x + 2  
end  
